Revert -out option checks

levitte opened the pull request:
The checks were over elaborate and became so increasingly the more we kept trying.

This reverses #4008 and #3709, i.e. reinstates that very light checks, but keeps `test/recipes/15-test_out_option.t`.  This puts us back at where we started with #3404 and allows us to deal with it differently that we have 'til now.  This also fixes #5590, and cancels out #5592 and #5638.

The first commit is from #6031

levitte committed:
test/recipes/15-test_out_option.t: add further tests

Test writing to the null device.  This should be successful.
Test writing to a writable file in a write protected directory.  This
should be successful.

Also, refactor so the planned number of tests is calculated.

[extended tests]

levitte committed:
Revert "Check on VMS as well"

This reverts commit f6d765988f37c43edb1056ab83165f2569182e9d.

levitte committed:
Revert "Add VMS version of app_dirname()"

This reverts commit 215a6730f1eaf53b01a4eb10d75bd09fd74f70cc.

levitte committed:
Revert "Check directory is able to create files for various -out option"

This reverts commit 555c94a0db9661428da0a45cb32b9f002324eefd.

levitte committed:
Better check of return values from app_isdir and app_access

[extended tests]

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

levitte commented:
@petrovr, I  know this has been a pet peeve.  Your comments would be appreciated.

richsalz reviewed (approved):
I think it's time to just walk away from this whole issue for this release, so this PR is good.

bernd-edlinger commented:
There appveyor failure seems relevant:

```
[00:12:33] Generating RSA private key, 16384 bit long modulus (2 primes)
[00:12:33] ...................................................................................................................++
[00:12:33] ....................................................................................................................++
[00:12:33] e is 65537 (0x010001)
[00:12:33] ..\..\apps\openssl.exe genrsa -out test_out_option-nowrite-2628\unwritable.pem 16384 => 0
[00:12:33] not ok 9 - invalid output path: test_out_option-nowrite-2628\unwritable.pem
[00:12:33] 
[00:12:33] #   Failed test 'invalid output path: test_out_option-nowrite-2628\unwritable.pem'
[00:12:33] #   at ..\test\recipes\15-test_out_option.t line 82.
[00:12:33] not ok 10 - check time consumed
[00:12:33] 
[00:12:33] #   Failed test 'check time consumed'
[00:12:33] #   at ..\test\recipes\15-test_out_option.t line 85.
```

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

levitte commented:
@bernd-edlinger, I've added a fixup that should help.

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

levitte commented:
Hmmm, is it just me, or does MSVC's `_access` not recognise `NUL`?  @dot-asm, you seem to know these things better than most, would you be willing to shed some light on this?

dot-asm commented:
> is it just me, or does MSVC's _access not recognise NUL?

Hmmm... It apparently varies with CRT version... I don't know that it changed, but latest CRT apparently returns EINVAL upon attempt to _access("NUL"). On the other hand default CRT, c:\windows\system32\msvcrt.dll, one mingw links with, acknowledges NUL's existence and recognizes it as writable... I suppose MSC's _access is not to be trusted...

levitte commented:
So basically, we (and any application, really) have to handle the magic file "nul" ourselves. That's kinda crappy... But ok, I can add that. Do we have to be equally attentive with the rest of the "POSIX API" (thinking of `_open` in particular)?

levitte commented:
Because the check of `-out` parameter calls `app_access` on its argument. I think we could remove that check, really, but haven't dared to, yet. Either way, we may have to handle "nul" specifically when opening the file...

bernd-edlinger commented:
and "con", "com1", etc...?
and "\\.\nul" as well?

levitte commented:
Good question, and I would rather not have to, so alternate solutions are welcome.
/notawindowsguy

bernd-edlinger commented:
Me neither....
I haven't looked at the code yet, but my feeling is we should probably remove the _access test,
maybe instead only check that the string is not empty, and be prepared that open can fail.

dot-asm commented:
> For example, you can chmod 0555 a directory on [Linux] vfat as much as you like, it won't make any difference...

Same can be said about Windows, either FAT or NTFS. chmod 0555 adds read-only attribute to directory, but it has no effect on your ability to create files in the directory.

levitte commented:
I wouldn't remove the test yet, I wanna see that our code can handle what the user throws at us correctly (i.e in a manner that a user should be able to expect, such as "nul" on the command prompt being correctly interpreted as the null device and not generating a silly error).
Considering `_access` doesn't seem to handle "nul" (i.e treats it as invalid), I'm questioning the whole "POSIX API" (including `_open`), and until we have shown for ourselves that we are treating this right, I think the recipe should stay. Perhaps not in its entirety, but still...

levitte committed:
apps/opt.c: Remove the access checks of input and output files

open() will take care of the checks anyway

dot-asm commented:
> I wouldn't remove the test yet,

Then for negative testing I'd say *one* directory is sufficient, ".", and test for file in non-existing directory. I'd also say that there is no need to test for time (it's unreliable anyway), just check for exit code would do.

> I wanna see that our code can handle what the user throws at us correctly (i.e in a manner that a user should be able to expect, such as "nul" on the command prompt being correctly interpreted as the null device and not generating a silly error).

Do [Windows] users actually have such specific expectations? Besides, you really can't do better than underlying system and systems do perform differently. So that there is no such thing as universal expectations, they have to be adjusted to specifics.

> I'm questioning the whole "POSIX API" (including _open)

_open can open "nul". Because it actually translates to native system call. But what "POSIX API"? Does POSIX specify "_access" or "_open"? No. It's "access" and "open". What Microsoft provides is a porting aid, not actual POSIX API. It has limitations, and user should be aware of them. Once again, you can't do better than underlying system.

levitte commented:
> But what "POSIX API"? Does POSIX specify "_access" or "_open"? No. It's "access" and "open".

Exactly, that's why I wrote POSIX API within quotes, because it isn't really.

> Do [Windows] users actually have such specific expectations?

I don't understand the question...  in Unix and on VMS, there is a null device, and users on those platforms expect whatever they send that way to disappear.  Are you telling me Windows users don't want to do something like that as well?

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

Reduce the amount of testing.  We only need to check that one attempt
to write to a directory fails, and we can skip checking writability to
"protected" directories.  Finally, we skip trying to check by timing,
and we can further reduce the testing code to make it look less
complex.

levitte commented:
I reduced the test recipe as you suggested.

dot-asm commented:
> Are you telling me Windows users don't want to do something like that as well?

What I'm saying is that they should be able to accept the fact that it doesn't work with openssl.exe. Well, it's sufficient to not use _access, but then question is this. If we accept the premise that access call constitutes a meaningful check, then what would you appreciate more, having it tell you about *all* regular files, or fail with just "nul"? Or in other words it's more important that it works with regular files, and users should be able to accept that "nul" doesn't work with openssl.exe, i.e. lower their expectations.

levitte commented:
> i.e. lower their expectations.

For something as well know as "nul"?  I think that's harsh, really.

But, what I've done right now is to remove the `access()` calls.  They are unnecessary anyway, as `open()` will redo the same checks at the time of opening, and if it is as you say, that `_open()` handles the magic files correctly, then everything will be fine.  So maybe we can wait with further arguments until the CIs have shown us what that gives?

dot-asm commented on test/recipes/15-test_out_option.t:
```diff
@@ -16,46 +16,51 @@ use OpenSSL::Test::Utils;
+
+foreach (@failure_paths) {
+    my $path = File::Spec->canonpath($_);
+    ok(!run(app([ 'openssl', 'genrsa', '-out', $path, '2048'])),
```
Does it have be such expensive operation? Of course it's hardly noticeable on your desktop, but do you have to make Raspberry Pi user suffer [more than necessary]? rand -out $path would do... Or why not convert something from pem to der?

dot-asm replied:
Well, I suppose it applies more to success tests. That is if we assume that failure test would fail early....

levitte replied:
The reason for it is that this was already in the recipe before I started fiddling with it, and I simply didn't think further.  Of course we can change that to something better suited for quick checks...

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

use "openssl rand" rather than "openssl genrsa"

dot-asm commented on test/recipes/15-test_out_option.t:
```diff
@@ -16,46 +16,51 @@ use OpenSSL::Test::Utils;
+# For example, if cross compiled and tested on the build host, perl will
+# generate an incorrect NULL device name.
+# We might expand the exceptions...
+unless (config('target') =~ m|^mingw| && $^O ne 'msys') {
```
Even though it's possible to test Linux cross-compiled targets with qemu-user (since they use same file system), it's not really inappropriate to simply assume that perl's devnull has no meaning to *any* cross-compiled target. Incidentally this will work out even for mingw builds performed under Linux and Cygwin. Because they are customarily performed with --cross-compile-prefix. In other words one can wonder if simple check for cross-compile prefix would be adequate. Do note "adequate", as this is really what it's about. I mean it's effectively impossible to specify condition that is guaranteed to work in all circumstances, you can only go for what has best chances to work in most cases. Of course it has to work with CIs, but as long as _access is out of picture check for cross-compile prefix should do the trick...

levitte replied:
Oh, so now I *should* check `config('CROSS_COMPILE')`?  'cause [back over here](https://github.com/openssl/openssl/pull/5592#pullrequestreview-103906512), you told me a different story.

...  buuuuuuut ok, you say check cross compile prefix, I will check cross compile prefix.

dot-asm replied:
> Oh, so now I should check config('CROSS_COMPILE')?

I said "one can *wonder* if it would be adequate," which is not same say "do this" :-)

> 'cause back over here, you told me a different story.

But then I also said "Grrrr! MSYS2 cheats!" I.e. by that time I didn't actually realize that msys perl would convert "/dev/null" to "nul" when starting native binary such as "mingw". I thought that it would do same as Linux and Cygwin perls, i.e. actually pass "/dev/null".

dot-asm replied:
> "Grrrr! MSYS2 cheats!"

Besides, context is different now. That's what makes it hard to nail, which can be frustrating and appear contradicting at different occasions. It's multi-variable problem with no exact solution. Change one variable, like _access call removal, and suddenly check for cross-compile prefix makes more sense than other possible options. But it's just that it makes *more* sense, not that it's absolutely "the right thing to do ever"...

levitte replied:
> I said "one can wonder if it would be adequate," which is not same [as to] say "do this" :-)

True that...  well, it *does* cover the mingw on Linux cases, and I've been wondering more than once what happens with other explicitly cross compiled environments.  I think this covers quite a lot of the cases we commonly test for, though...  right?

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

dot-asm commented:
As soon as WIP is removed...

mspncp commented:
>> i.e. lower their expectations.

>For something as well know as "nul"?  I think that's harsh, really.

I don't think it's necessary to lower their expectations. Peoply which are stuck with cmd.exe generally have _very low_ expectations about their shell. ;-)

levitte commented:
WIP removed.

dot-asm commented on test/recipes/15-test_out_option.t:
```diff
@@ -16,46 +16,49 @@ use OpenSSL::Test::Utils;
+    './',
+   );
+my @success_paths = (
+    'test.pem'
```
Now with rand -out this is no longer suitable name. test.bin perhaps? Or random.bin? Or randomly generated name with .bin extension?

levitte replied:
Ah, thanks for the reminder, that was on my mind before going to sleep yesterday

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

dot-asm reviewed (approved):
Red cross from appveyor is unrelated.

levitte commented:
Merged (including commit from #6031)

4522e130c87c341342c640bba970f4b89755f1cb apps/opt.c: Remove the access checks of input and output files
c36e9093914aab4bfc42af1db35558a9272607b5 Better check of return values from app_isdir and app_access
96de2e590bad00575baa7c2c6be5767b43aa017c Revert "Check directory is able to create files for various -out option"
b47b6650164711de449a7738112a3efb733213e2 Revert "Add VMS version of app_dirname()"
b9a354d08810e95c03dec1f796fdd4069a4001e7 Revert "Check on VMS as well"
39e32be1ccc527036e0a52c3cf64a40faf28ab9d test/recipes/15-test_out_option.t: refine tests
